baselines/FedPer/FedPer/implemented_models/cnn_model.py

```diff
@@ -0,0 +1,193 @@
+        # self.head = nn.Linear(84, 10)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass."""
+        x = self.body(x)
+        return self.head(x)
+
+
+class CNNModelSplit(ModelSplit):
+    """Split the CNN model into body and head."""
+
+    def _get_model_parts(self, model: CNNNet) -> Tuple[nn.Module, nn.Module]:
+        return model.body, model.head
+
+
+class CNNModelManager(ModelManager):
+    """Manager for models with Body/Head split."""
+
+    def __init__(
+        self,
+        client_id: int,
+        config: Dict[str, Any],
+        trainloader: DataLoader,
+        testloader: DataLoader,
+        has_fixed_head: bool = False,
+    ):
+        """Initialize the attributes of the model manager.
+
+        Args:
+            client_id: The id of the client.
+            config: Dict containing the configurations to be used by the manager.
+            has_fixed_head: Whether a fixed head should be created.
+        """
+        super().__init__(
+            model_split_class=CNNModelSplit,
+            client_id=client_id,
+            config=config,
+            has_fixed_head=has_fixed_head,
+        )
+        self.trainloader, self.testloader = trainloader, testloader
+        self.device = self.config["device"]
+
+    def _create_model(self) -> nn.Module:
+        """Return CNN model to be splitted into head and body."""
+        try:
+            return CNNNet(name="cnn").to(self.device)
+        except AttributeError:
+            self.device = self.config["device"]
+            return CNNNet(name="cnn").to(self.device)
+
+    def train(
+        self,
+        train_id: int,
+        epochs: int = 1,
+        tag: Optional[str] = None,
+        fine_tuning: bool = False,
+    ) -> Dict[str, Union[List[Dict[str, float]], int, float]]:
+        """Train the model maintained in self.model.
+
+        Method adapted from simple CNN from Flower 'Quickstart PyTorch' \
+        (https://flower.dev/docs/quickstart-pytorch.html).
+
+        Args:
+            train_id: id of the train round.
+            epochs: number of training epochs.
+            tag: str of the form <Algorithm>_<model_train_part>.
+                <Algorithm> - indicates the federated algorithm that is being performed\
+                              (FedAvg, FedPer).
+                <model_train_part> - indicates the part of the model that is
+                being trained (full, body, head). This tag can be ignored if no
+                difference in train behaviour is desired between federated algortihms.
+            fine_tuning: whether the training performed is for model fine-tuning or not.
+
+        Returns
+        -------
+            Dict containing the train metrics.
+        """
+        criterion = torch.nn.CrossEntropyLoss()
+        optimizer = torch.optim.SGD(self.model.parameters(), lr=0.001, momentum=0.9)
+        for _ in range(epochs):
+            for images, labels in tqdm(self.trainloader):
```

Comment:
Unless you think it's beneficial, i'd vote for removing `tqdm` (it makes the log unnecessarily long). I think it's fine to keep it for centralised evaluation on the server side if you implement it. But having this on the client side is not useful the moment there are more than one client training concurrently.